bin/main.rb

```diff
@@ -1,94 +1,89 @@
-    # Show available numbers from the updated array of options
-    selection = gets.chomp
-    selection = selection.to_i
-    # moves_counter == 0 only checks if it is a number, not if is an available cell, else
-    # make mentioned checks, if ok, assings value to cell position else repeat
-    # remove number choice from array of options
-    # check if player has made a win move,
-    if moves_counter > 4
-      puts "#{first_player} has won! congrats!!" + '--> if there\'s a winning move, of course, now breaks'
-      # somebody_won = true
-      # breaks the loop
-      # check if have been reached 9 moves and no winner,
-      puts 'there\'s been a tie' + '--> if there is a tie, of course, now breaks'
-      # break the  loop
-    end
-    puts ' '
+# links to classes inside /lib
+
+require_relative '../lib/board'
+require_relative '../lib/player'
+
+# Introduce the game to the players
+
+def cover_image
+  puts
+  puts '           ▛┅┅┅┳┅┅┅┳┅┅▜'
+  puts '           ┇ 1 ┇ 2 ┇ 3 ┇'
+  puts '           ┣┅┅┅╋┅┅┅╋┅┅┅┫'
+  puts '           ┇ 4 ┇ 5 ┇ 6 ┇'
+  puts '           ┣┅┅┅╋┅┅┅╋┅┅┅┫'
+  puts '           ┇ 7 ┇ 8 ┇ 9 ┇'
+  puts '           ▙┅┅┅┻┅┅┅┻┅┅▟'
+  puts
+  puts 'Welcome to TIC-TAC-TOE by maosan132'
+  puts
+end
+
+game_on = true
+new_board = Board.new
+count = 0
+
+def update_screen
+  Gem.win_platform? ? (system 'cls') : (system 'clear')
+  cover_image
+end
+
+update_screen
+print 'Player one, please type your name: '
+player_name = gets.chomp
+player_one = Player.new(player_name)
+puts "Welcome, #{player_one.name}. The X is yours"
+
+print 'Player two, please type your name: '
+player_name = gets.chomp
+player_two = Player.new(player_name)
+puts "Welcome, #{player_two.name}. The O is yours"
+
+while game_on
+  update_screen
+  new_board.display
+  count += 1
+
+  if count.odd?
+    turn = 'X'
+    player = player_one.name
   else
-    puts "#{second_player}, make a choice"
-    # Show available numbers from the updated array of options
-    selection = gets.chomp
-    selection = selection.to_i
-    # make mentioned checks, if ok, assings value to cell position else repeat
-    # remove number choice from array of options
-    # if player has made a win move,
-    if moves_counter > 4
-      puts "#{second_player} has won! congrats!!" + '--> if there is a winning move, of course, now breaks'
-      # somebody_won = true
-      # breaks the loop
-      # check if have been reached 9 moves and no winner,
-      puts 'there is been a tie' + '--> if there\'s a tie, of course, now breaks'
-      # break the  loop
-    end
-    puts
+    turn = 'O'
+    player = player_two.name
+  end
+  puts "turn is #{turn}"
+  print "#{player}, Choose a cell number from 1 to 9: "
+  cell_choice = gets.chomp.to_i
+
+  valid = new_board.valid(cell_choice)
+  unless valid
+    print "#{turn}, Choose a cell number from 1 to 9: "
+    cell_choice = gets.chomp.to_i
   end
-  # displays the updated board
-  puts '<-----Draws the updated board----->'
+
```

Comment:
You can leave this section without a valid choice.
You must loop until obtaining a valid selected cell. It's a must!

Reply:
I appreciate your help, but I disagree here. The code indeed is looping: it checks every input until it satisfies the **valid** method (what you are pointing out), otherwise outputs again and again: "Choose a number between 1 and 9".